cmake/gtest_utils.cmake

```diff
@@ -0,0 +1,75 @@
+enable_testing()
+
+function(k2_add_gtest)
+  cmake_parse_arguments(
+      ARGS                                 # prefix of output variables
+      "NO_MAIN;NO_GMOCK;EXCLUDE_FROM_ALL"  # list of names of the boolean arguments (only defined ones will be true)
+      ""                                   # list of names of mono-valued arguments
+      "SOURCES;DEPENDS"                    # list of names of multi-valued arguments (output variables are lists)
+      ${ARGN}                              # arguments of the function to parse, here we take the all original ones
+  ) # remaining unparsed arguments can be found in ARGS_UNPARSED_ARGUMENTS
+
+  set(dependencies ${ARGS_DEPENDS})
+
+  list(LENGTH ARGS_UNPARSED_ARGUMENTS other_args_size)
+  if(other_args_size EQUAL 1)
+    set(name ${ARGS_UNPARSED_ARGUMENTS})
+  else()
+    message(FATAL_ERROR "too many or not enough unnamed arguments")
+  endif()
+
+  if(ARGS_SOURCES)
+    set(sources ${ARGS_SOURCES})
+  else()
+    message(FATAL_ERROR "you must specify at least one source for the test")
+  endif()
+
+  if(ARGS_EXCLUDE_FROM_ALL)
+    set(EXCLUDE_FROM_ALL "EXCLUDE_FROM_ALL")
+  endif()
+
+  string(REGEX REPLACE "^.*\\." "" tgt ${name})
+  add_executable(${tgt} ${sources} ${EXCLUDE_FROM_ALL})
+
+  if(ARGS_NO_GMOCK)
+    target_link_libraries(${tgt} PUBLIC ${dependencies} gtest)
```

Comment:
why is it `PUBLIC` for `gtest` and `gmock`, but `private` for the subsequent `gtest_main` and `gmock_main`?

Reply:
The PRIVATE keep the most depends on the target. It's suggested to compile executables.

Reply:
_main would lead to an executable.